st2common/st2common/models/api/execution.py

```diff
@@ -136,7 +136,10 @@ def to_model(cls, instance):
             if not value and not cls.model._fields[attr].required:
                 continue
             if attr not in ActionExecutionAPI.SKIP:
-                setattr(model, attr, value)
-        model.start_timestamp = isotime.parse(instance.start_timestamp)
-        model.end_timestamp = isotime.parse(instance.end_timestamp)
+                values[attr] = value
+
+        values['start_timestamp'] = isotime.parse(instance.start_timestamp)
+        values['end_timestamp'] = isotime.parse(instance.end_timestamp)
+
+        model = cls.model(**values)
```

Comment:
@manasdk I posted an explanation on Slack two days ago and there is some in 553c57f1c9de9934867a34fdd2069d644b1d2b1a.

In short, previously we had no idea when (and if) the variables were available and this made constructor useless. If we wanted to, for example, generate and assign a derived field based on values of other field values, we couldn't do that consistently. We could, for example, do that in `save()`, but that's mongoengine specific and that field would only be available after save is called.

As far as "expensivness" goes, if you are referring to a temporary dict here and temporary variables in other places, then this is a non-issue. We have 999 other places where performance might eventually become an issue (logging, queries, etc.), but this is not one of them.